selogin module

jamescassell opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Move selogin module upstream from linux-system-roles/selinux project.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes: https://github.com/linux-system-roles/selinux/issues/34

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
selogin

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

bachradsusi committed:
selogin

original commit:
https://github.com/linux-system-roles/selinux/commit/8efa4cdde9a9cc2adaf487e96d9e93b7d7bcf8c6

update module to match current ansible standards

- update version_added
- fix sanity checks
- match docs to module definition of required and default values
  - explicit default of 's0' for selevel
  - no default for seuser or login
- rename serange->selevel, add alias
- add required imports
- add RETURN info (stolen from sister module)
- add author github handles
- remove star-import, name specific requirements
- handle import error similar to sister modules
- collapse exceptions
- remove IOError, alias for OSError
  - https://stackoverflow.com/questions/29347790/difference-between-ioerror-and-oserror
- drop 'required: false' as false is the default
- list dependencies as python-agnostic
- selogin does not support check_mode
- remove unused functions
- require seuser only if state=present

Co-authored-by: James Cassell <code@james.cassell.me>

jamescassell commented:
@dankeder @bachradsusi could you take a look and make sure I haven't messed anything up wrt module defaults, and that module argument names make sense?  I tried to match them to the surrounding SELinux modules.

Happy to address any feedback.  If something should be renamed, let me know.

maxamillion commented on lib/ansible/modules/system/selogin.py:
```diff
@@ -0,0 +1,299 @@
+            state=dict(type='str', default='present', choices=['absent', 'present']),
+            reload=dict(type='bool', default=True),
+        ),
+        supports_check_mode=True
```
The module claims to support check mode but never actually checks if the module is being executed in check mode.

jamescassell replied:
That's what reviews are for! Thanks for catching that!

jamescassell replied:
Removed claim of supporting check_mode.  I may add check_mode support if time permits, but I'd really like to get it merged for 2.8 with or w/o check_mode.

jamescassell replied:
I've added support for check_mode

maxamillion commented:
The code looks good for the most part, I'd like to see the `check_mode` comment I made addressed and integration tests added. You can use https://github.com/ansible/ansible/tree/devel/test/integration/targets/selinux as an example. Thank you!

bachradsusi commented on lib/ansible/modules/system/selogin.py:
```diff
@@ -0,0 +1,299 @@
+from ansible.module_utils.basic import AnsibleModule, missing_required_lib
+
+
+def semanage_port_get_ports(seport, setype, proto):
```
This function and the function bellow (`semanage_port_get_type`) are probbaly leftovers from the seport.py module. Neither of them is used in this module and ports have no relation to SELinux login mappings.

jamescassell replied:
removed

bachradsusi commented on lib/ansible/modules/system/selogin.py:
```diff
@@ -0,0 +1,299 @@
+notes:
+   - The changes are persistent across reboots
+   - Not tested on any debian based system
+requirements: [ 'libselinux-python', 'policycoreutils-python' ]
```
This was originally written for Red Hat Enterprise Linux 7. But there's no libselinux and policycoreutils python2 bindings in Red Hat Enterprise Linux 8 Beta, and they'll be dropped in Fedora 31 as well. The same comment can be applied to lines 266 and 269

jamescassell replied:
Are there replacement libraries we should mention here? (Or is it just `-python3`?)

bachradsusi replied:
Python 3 versions are packaged in `python3-libselinux` and `python3-policycoreutils`. But the module needs just `selinux` and `seobject` python modules and version depends on the current interpret version. I don't know how to make it understandable and correct.

On lines 267 and 270 I'd probably use module names could be used instead of package names:

````
module.fail_json(msg=missing_required_lib("selinux"), exception=SELINUX_IMP_ERR)
module.fail_json(msg=missing_required_lib("seobject"), exception=SELINUX_IMP_ERR)
````

jamescassell replied:
I changed it to `libselinux` and `seobject provided by policycoreutils`

bachradsusi commented on lib/ansible/modules/system/selogin.py:
```diff
@@ -0,0 +1,299 @@
+            selogin.add(login, seuser, serange)
+            change = True
+        else:
+            selogin.modify(login, seuser, serange)
```
Missing `change = True` ?

jamescassell replied:
I think it was not present because otherwise the module would have always reported "changed" because it didn't check whether a change was made.  It only reported a change if a login mapping was added, not modified.

In the latest commit, I've added check_mode support, which also fixes this issue.

jamescassell committed:
selogin: check_mode, ignore_selinux_state, tests

ignore_selinux_state based on #48945

jamescassell commented:
ready_for_review

@maxamillion @bachradsusi thanks for the reviews!  I believe I've fully addressed your feedback.  Could you take another look when you get a chance?

I've added integration tests and support for `check_mode` and `ignore_selinux_state`

Hoping to get this merged for 2.8.

maxamillion commented:
rebuild_merge

maxamillion commented:
rebuild_merge

maxamillion commented:
rebuild_merge
